chacha20/src/block/mod.rs

```diff
@@ -45,66 +45,28 @@ impl Block {
             ],
         };
 
-        block.rounds();
-        block.finish(key, iv, counter)
+        block.rounds(20)
     }
 
     /// Run the 20 rounds (i.e. 10 double rounds) of ChaCha20
     #[inline]
-    fn rounds(&mut self) {
-        self.double_round();
-        self.double_round();
-        self.double_round();
-        self.double_round();
-        self.double_round();
-
-        self.double_round();
-        self.double_round();
-        self.double_round();
-        self.double_round();
-        self.double_round();
-    }
-
-    /// Double round function
-    #[inline]
-    fn double_round(&mut self) {
-        let state = &mut self.state;
-        quarter_round(0, 4, 8, 12, state);
-        quarter_round(1, 5, 9, 13, state);
-        quarter_round(2, 6, 10, 14, state);
-        quarter_round(3, 7, 11, 15, state);
-        quarter_round(0, 5, 10, 15, state);
-        quarter_round(1, 6, 11, 12, state);
-        quarter_round(2, 7, 8, 13, state);
-        quarter_round(3, 4, 9, 14, state);
-    }
-
-    /// Finish computing a block
-    #[inline]
-    fn finish(
-        self,
-        key: &[u32; KEY_WORDS],
-        iv: [u32; IV_WORDS],
-        counter: u64,
-    ) -> [u32; STATE_WORDS] {
+    fn rounds(&mut self, count: usize) -> [u32; STATE_WORDS] {
```

Comment:
Additionally this parameterizes the number of rounds enabling the implementation of ChaCha8 and ChaCha12. Again, this comes at no cost to performance (or constant-time, as the number of rounds is a constant).